src/rxzsets.c

```diff
@@ -304,19 +308,27 @@ int ZUnionTopKCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
   }
   Make_Heap(v, 0, v->top, rev ? __zsetentry_less : __zsetentry_greater);
 
+  khash_t(32) *h = kh_init(32);
   size_t reply_count = 0;
   RedisModule_ReplyWithArray(ctx, REDISMODULE_POSTPONED_ARRAY_LEN);
-  for (size_t i = 0; i < k && v->top != 0; i++) {
+  while (kh_size(h) < k && v->top != 0) {
     // pop from heap
     Heap_Pop(v, 0, v->top, rev ? __zsetentry_less : __zsetentry_greater);
     ZsetEntry *entry = (ZsetEntry *) (v->data + ((v->top - 1) * v->elemSize));
-    // reply to client
-    RedisModule_ReplyWithString(ctx, entry->element);
-    reply_count++;
-    if (with_scores) {
-      RedisModule_ReplyWithDouble(ctx, entry->score * entry->weight);
+
+    // de-duplicate
+    int ret = 0;
+    kh_put(32, h, RedisModule_StringPtrLen(entry->element, NULL), &ret);
```

Comment:
this might be dangerous re NULL termination. 
from client requests strigns are null terminated implicitly, but from RedisModule_Call responses they usually aren't. 
You can `strndup` maybe and create a null terminated string if you can free it later.